miriway.cpp

```diff
@@ -273,9 +294,63 @@ int main(int argc, char const* argv[])
 
     // A configuration option to start applications when compositor starts and record them in `shell_pids`.
     // Because of the previous section, this allows them some extra Wayland extensions
-    std::function<void(std::vector<std::string> const&)> shell_launch = [&](std::vector<std::string> const& cmd)
+    std::function<void(std::vector<std::string> const&, ShellComponentRunInfo&)> shell_launch
+        = [&](std::vector<std::string> const& cmd, ShellComponentRunInfo& info)
         {
-            shell_pids.insert(client_launcher.launch(cmd), [&shell_launch, cmd] { shell_launch(cmd); });
+            time_t now;
+            time(&now);
+            auto const time_elapsed_since_last_run = now - info.last_run_time;
+            if (time_elapsed_since_last_run <= min_time_seconds_allowed_between_runs)
+            {
+                // The command is trying to restart too soon, so we throttle it. For every premature
+                // death that it has in a row, we increasingly throttle its restart time.
+                // After MAX_TOO_SHORT_RERUNS_IN_A_ROW is hit, we stop trying to rerun
+                if (info.num_runs_with_too_short_time_in_between >= max_too_short_reruns_in_a_row)
+                {
+                    mir::log_warning("No longer restarting app: %s", get_cmd_string(cmd).c_str());
+                    return;
+                }
+
+                auto const time_to_wait = time_seconds_wait_foreach_run * (info.num_runs_with_too_short_time_in_between + 1);
+                auto const timer_fd = timerfd_create(CLOCK_REALTIME, 0);
+                if (timer_fd == -1)
+                {
+                    mir::log_error("timerfd_create failed, unable to restart application: %s", get_cmd_string(cmd).c_str());
+                    return;
+                }
+
+                auto const spec = itimerspec
+                {
+                    { 0, 0 },               // Timer interval
+                    { time_to_wait, 0 } // Initial expiration
+                };
+
+                if (timerfd_settime(timer_fd, 0, &spec, NULL) == -1)
+                {
+                    mir::log_error("timerfd_settime failed, unable to restart application: %s", get_cmd_string(cmd).c_str());
+                    return;
+                }
+
+                info.handle = runner.register_fd_handler(mir::Fd{timer_fd}, [&](int){
+                    time_t now;
+                    time(&now);
+                    info.handle.reset();
+                    info.num_runs_with_too_short_time_in_between++;
+                    info.last_run_time = now;
+                    shell_pids.insert(client_launcher.launch(cmd), [&shell_launch, &info, &cmd] {
+                        shell_launch(cmd, info);
+                    });
```

Comment:
This is prone to undefined behaviour: `&info` is a pointer to an element of a vector. And if, before the lambda executes, the vector reallocates then the pointer will have been invalidated.

Why not capture `info` by value? Then the problematic vector can be avoided.

Reply:
Yeah I see what you mean. I can just keep calling it with a `ShellComponentRunInfo` instead of a `ShellComponentRunInfo&`. I _think_ that will work